### PR TITLE
Time for layerClusters in HGCAL

### PR DESCRIPTION
**Pull request by amartelli**
produce a ValueMap with the computed time for 2d clusters, so the BasicCluster dataformat is not modified

### CODE REVIEW

**Comment by slava77** [comment]
@cmsbuild please test

File: RecoLocalCalo/HGCalRecProducers/plugins/HGCalLayerClusterProducer.cc [hunk 1]
```diff
@@ -147,25 +159,38 @@ void HGCalLayerClusterProducer::produce(edm::Event& evt,
...
+  //make a map detid-rechit
+  // NB for the moment just host EE and FH hits
+  // timing in digi for BH not implemented for now
+  std::map<const HGCalDetId, float> hitmap;
+
   switch(algoId){
   case reco::CaloCluster::hgcal_em:
     evt.getByToken(hits_ee_token,ee_hits);
     algo->populate(*ee_hits);
+    for(auto it: *ee_hits) hitmap[it.detid()] = it.time();
```
Comment (slava77) [inline RecoLocalCalo/HGCalRecProducers/plugins/HGCalLayerClusterProducer.cc]: ```auto const&```

File: RecoLocalCalo/HGCalRecProducers/plugins/HGCalLayerClusterProducer.cc [hunk 2]
```diff
@@ -147,25 +159,38 @@ void HGCalLayerClusterProducer::produce(edm::Event& evt,
...
     break;
   case  reco::CaloCluster::hgcal_had:
     evt.getByToken(hits_fh_token,fh_hits);
     evt.getByToken(hits_bh_token,bh_hits);
     if( fh_hits.isValid() ) {
       algo->populate(*fh_hits);
+      for(auto it: *fh_hits) hitmap[it.detid()] = it.time();
```
Comment (slava77) [inline RecoLocalCalo/HGCalRecProducers/plugins/HGCalLayerClusterProducer.cc]: ```auto const&```

File: RecoLocalCalo/HGCalRecProducers/plugins/HGCalLayerClusterProducer.cc [hunk 3]
```diff
@@ -147,25 +159,38 @@ void HGCalLayerClusterProducer::produce(edm::Event& evt,
...
     } else if ( bh_hits.isValid() ) {
       algo->populate(*bh_hits);
     }
     break;
   case reco::CaloCluster::hgcal_mixed:
     evt.getByToken(hits_ee_token,ee_hits);
     algo->populate(*ee_hits);
+    for(auto it: *ee_hits){
```
Comment (slava77) [inline RecoLocalCalo/HGCalRecProducers/plugins/HGCalLayerClusterProducer.cc]: ```auto const&```

File: RecoLocalCalo/HGCalRecProducers/plugins/HGCalLayerClusterProducer.cc [hunk 4]
```diff
@@ -147,25 +159,38 @@ void HGCalLayerClusterProducer::produce(edm::Event& evt,
...
+      hitmap[it.detid()] = it.time();
+    }
     evt.getByToken(hits_fh_token,fh_hits);
     algo->populate(*fh_hits);
+    for(auto it: *fh_hits){
```
Comment (slava77) [inline RecoLocalCalo/HGCalRecProducers/plugins/HGCalLayerClusterProducer.cc]: ```auto const&```

File: RecoLocalCalo/HGCalRecProducers/plugins/HGCalLayerClusterProducer.cc [hunk 5]
```diff
@@ -180,11 +205,39 @@ void HGCalLayerClusterProducer::produce(edm::Event& evt,
...
   for( unsigned i = 0; i < clusterHandle->size(); ++i ) {
     edm::Ptr<reco::BasicCluster> ptr(clusterHandle,i);
     clusterPtrs.push_back(ptr);
+
+    float timeCl = -99.;
+    const reco::CaloCluster &sCl = (*clusterHandle)[i];
+    if(sCl.size() >= 3){
+      std::vector<float> timeClhits;
+
+      for(auto hit : sCl.hitsAndFractions()){
```
Comment (slava77) [inline RecoLocalCalo/HGCalRecProducers/plugins/HGCalLayerClusterProducer.cc]: ```auto const&```

File: RecoLocalCalo/HGCalRecProducers/plugins/HGCalLayerClusterProducer.cc [hunk 6]
```diff
@@ -180,11 +205,39 @@ void HGCalLayerClusterProducer::produce(edm::Event& evt,
...
+        std::map<const HGCalDetId, float>::iterator finder = hitmap.find(hit.first);
```
Comment (slava77) [inline RecoLocalCalo/HGCalRecProducers/plugins/HGCalLayerClusterProducer.cc]: ```const_iterator ``` or ```auto```

File: RecoLocalCalo/HGCalRecProducers/plugins/HGCalLayerClusterProducer.cc [hunk 7]
```diff
@@ -147,25 +159,38 @@ void HGCalLayerClusterProducer::produce(edm::Event& evt,
 
   algo->getEventSetup(es);
 
+  //make a map detid-rechit
+  // NB for the moment just host EE and FH hits
+  // timing in digi for BH not implemented for now
+  std::map<const HGCalDetId, float> hitmap;
```
Comment (kpedro88) [inline RecoLocalCalo/HGCalRecProducers/plugins/HGCalLayerClusterProducer.cc]: probably worthwhile to use an `unordered_map` here

File: RecoLocalCalo/HGCalRecProducers/plugins/HGCalLayerClusterProducer.cc [hunk 8]
```diff
@@ -180,11 +205,39 @@ void HGCalLayerClusterProducer::produce(edm::Event& evt,
...
+        float rhTime = finder->second;
+        if(rhTime < 0.) continue;
+        timeClhits.push_back(rhTime - timeOffset);
+      }
+      if(timeClhits.size() >= 3) timeCl = hgcalsimclustertime::fixSizeHighestDensity(timeClhits);
+    }
+    times.push_back(timeCl);
   }
 
+  std::unique_ptr<edm::ValueMap<float>> timeCl(new edm::ValueMap<float>());
```
Comment (kpedro88) [inline RecoLocalCalo/HGCalRecProducers/plugins/HGCalLayerClusterProducer.cc]: `auto timeCl = std::make_unique<edm::ValueMap<float>>();`

**Comment by perrotta** [comment]
please test

File: RecoLocalCalo/HGCalRecProducers/plugins/HGCalLayerClusterProducer.cc [hunk 9]
```diff
@@ -147,25 +159,38 @@ void HGCalLayerClusterProducer::produce(edm::Event& evt,
...
+  //make a map detid-rechit
+  // NB for the moment just host EE and FH hits
+  // timing in digi for BH not implemented for now
+  std::unordered_map<uint32_t, float> hitmap;
+
   switch(algoId){
   case reco::CaloCluster::hgcal_em:
     evt.getByToken(hits_ee_token,ee_hits);
     algo->populate(*ee_hits);
+    for(auto const& it: *ee_hits) hitmap[HGCalDetId(it.detid())] = it.time();
```
Comment (kpedro88) [inline RecoLocalCalo/HGCalRecProducers/plugins/HGCalLayerClusterProducer.cc]: I do not think the conversions to `HGCalDetId` are necessary or desirable. Just use `it.detid().rawId()` to get the raw value.

**Comment by kpedro88** [comment]
please test

**Comment by kpedro88** [comment]
@amartelli I looked at one of the output ROOT files from the matrix tests. The ValueMap is in the event content as expected. 95% of entries have the -99 default value, while the remaining 5% have values mostly between -1 and 1 with a small tail beyond 1. Is this the expected behavior?

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/24838	477690f 
- code changes are in line with the PR description
- jenkins tests pass and comparisons with the baseline (of previously known quantities) show no differences
- based on a local copy of the jenkins outputs shows that the new valueMap hgcalLayerClusters_timeLayerCluster shows up in outputs of the 2023 workflows and the values are filled with non-default value for a fraction of the clusters (the code fills non-default values only for 3+ measurements in a cluster; most data in the jenkins tests is noise and should have small clusters)
e.g. wf 20034 D17
all_baselinevs24838-30963_ttbar14tev2023d17wf20034p0c_min15 max-2 floatedmvaluemap_hgcallayerclusters_timelayercluster_reco_obj_values_
or with the default suppressed
all_baselinevs24838-30963_ttbar14tev2023d17wf20034p0c_floatedmvaluemap_hgcallayerclusters_timelayercluster_reco_obj_values_45

**Comment by kpedro88** [comment]
+upgrade
@amartelli please confirm the output that @slava77 and I have observed is correct

**Comment by fabiocos** [comment]
+1

**Comment by amartelli** [comment]
@kpedro88 yes what is observed is correct. The non default values are indeed filled for 3+ measurements with time in a cluster. sorry for the delay and thanks

**Pull request merged**